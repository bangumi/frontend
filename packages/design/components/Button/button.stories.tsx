import type { Meta, StoryFn } from '@storybook/react';
import type { FC } from 'react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { ArrowRightCircle } from '@bangumi/icons';

import type { ButtonLinkProps, ButtonProps } from '.';
import Button from '.';

const storyMeta: Meta<typeof Button> = {
  title: 'modern/Button',
  component: Button,
  subcomponents: { 'Button.Link': Button.Link as FC<unknown> },
  argTypes: {
    disabled: { control: 'boolean' },
    color: { control: 'select', options: ['default', 'blue', 'gray'] },
  },
  parameters: {
    docs: {
      description: {
        component:
          '主类型按钮，全圆角，有颜色填充，为默认类型。默认为粉色，另有蓝色和灰色两种颜色可选。',
      },
    },
  },
};

export default storyMeta;

const Template: StoryFn<ButtonProps> = (args) => {
  return <Button {...args}>{args.children ?? 'Click Me!'}</Button>;
};

export const Primary = Template.bind({});
Primary.args = {
  type: 'primary',
  children: '登录',
};

export const Secondary = Template.bind({});
Secondary.args = {
  type: 'secondary',
  children: '小组概览',
  size: 'medium',
};
Secondary.parameters = {
  docs: {
    description: {
      story:
        '次级按钮，全圆角，有边框，无颜色填充。默认为灰色，另有蓝色变种可选，此处为 `medium` 大小。',
    },
  },
};

export const Text = Template.bind({});
Text.args = {
  type: 'text',
  children: '编辑',
  size: 'small',
};
Text.parameters = {
  docs: {
    description: {
      story: '纯文字按钮，无边框，无颜色填充。默认为灰色，无其他颜色变种，此处为 `small` 大小。',
    },
  },
};

export const Blue = Template.bind({});
Blue.args = {
  type: 'primary',
  children: '发表新主题',
  color: 'blue',
};
Blue.parameters = {
  docs: {
    description: {
      story: '主类型按钮的蓝色变种，为默认的 `large` 大小。',
    },
  },
};

export const Small = Template.bind({});
Small.args = {
  type: 'secondary',
  children: '评论',
  size: 'small',
};
Small.parameters = {
  docs: {
    description: {
      story: '所有类型的按钮均提供 `large`、`medium`、`small` 三种尺寸。默认为 `large`。',
    },
  },
};

export const Square = Template.bind({});
Square.args = {
  shape: 'square',
};

export const Disabled = Template.bind({});
Disabled.args = {
  type: 'primary',
  disabled: true,
  children: 'Disabled',
};

const ButtonLinkTemplate: StoryFn<ButtonLinkProps> = (args) => {
  return (
    <MemoryRouter>
      <Button.Link {...args}>{args.children}</Button.Link>
    </MemoryRouter>
  );
};

export const Link = ButtonLinkTemplate.bind({});
Link.args = {
  to: 'https://bgm.tv',
  isExternal: true,
  target: '_blank',
  children: 'Bangumi 番组计划',
};
Link.parameters = {
  docs: {
    description: {
      story: '按钮的链接版本，使用方法与 `Typography.Link` 相同。',
    },
  },
};

export const Plain = ButtonLinkTemplate.bind({});
Plain.args = {
  to: 'https://bgm.tv',
  isExternal: true,
  target: '_blank',
  type: 'plain',
  children: (
    <>
      更多小组成员
      <ArrowRightCircle />
    </>
  ),
};
Plain.parameters = {
  docs: {
    description: {
      story: '无内边距、边框及圆角，最小高度。',
    },
  },
};
