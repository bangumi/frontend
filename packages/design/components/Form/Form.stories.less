.divider {
  height: 100%;
  margin: 0 14px;
}
