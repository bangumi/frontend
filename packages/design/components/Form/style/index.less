@import '../../../theme/base';

.bgm-form {
  display: inline-flex;
  flex-direction: column;
  gap: 1rem;

  &-item {
    display: flex;

    @media (max-width: @sm) {
      flex-direction: column;
    }

    &-label {
      display: inline-block;
      font-weight: 400;
      font-size: 1.125rem;
      line-height: 25px;
      color: @gray-60;

      > label {
        display: flex;
        align-items: center;
        height: 38px;
      }
    }
  }

  &--compact {
    gap: 0;
  }
}

// 点状分隔线，距离左右各 12px，故使用伪元素实现
.divider-pseudo-element() {
  &::after {
    content: '';
    position: absolute;
    bottom: -1px;
    width: calc(100% - 24px);
    border-bottom: 2px dotted @gray-10;
  }
}

.compact-form-item() {
  &:not(:first-child) {
    .no-top-border();
  }

  &:not(:last-child) {
    .no-bottom-border();
    .divider-pseudo-element();
  }
}
