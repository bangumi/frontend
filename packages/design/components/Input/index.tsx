import React, { forwardRef } from 'react';
import classnames from 'classnames';

export interface InputProps {
  /* 同原生 input 标签 `type` */
  type?: string;
  /* 同原生 `placeholder` */
  placeholder?: string;
  /* 外层 wrapper 的样式 */
  wrapperStyle?: React.CSSProperties;
  /* 外层 wrapper 的自定义类名 */
  wrapperClass?: string;
  /* 前缀 */
  prefix?: React.ReactNode;
  /* 后缀 */
  suffix?: React.ReactNode;
}

/* eslint-disable react/prop-types */
// https://github.com/jsx-eslint/eslint-plugin-react/issues/3140
const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ type = 'text', wrapperStyle, wrapperClass, prefix, suffix, placeholder, ...rest }, ref) => {
    return (
      <div className={classnames('bgm-input__wrapper', wrapperClass)} style={wrapperStyle}>
        {prefix && prefix}
        <input type={type} className='bgm-input' placeholder={placeholder} ref={ref} {...rest} />
        {suffix && suffix}
      </div>
    );
  },
);

export default Input;
