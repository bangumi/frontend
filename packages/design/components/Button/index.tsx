import './style';

import classNames from 'classnames';
import type { MouseEventHandler } from 'react';
import React from 'react';

import type { LinkProps } from '../Typography';
import Typography from '../Typography';

export interface ButtonCommonProps {
  type?: 'primary' | 'secondary' | 'text' | 'plain';
  shape?: 'square' | 'rounded';
  size?: 'large' | 'medium' | 'small';
  color?: 'default' | 'blue' | 'gray';
  /** 无内边距、边框及圆角，最小高度 */
}

export type ButtonProps = Omit<JSX.IntrinsicElements['button'], 'type' | 'onClick'> &
  ButtonCommonProps & {
    onClick?: MouseEventHandler; // desserts for story book
    htmlType?: JSX.IntrinsicElements['button']['type'];
  };

const Button = ({
  disabled = false,
  className,
  type = 'primary',
  shape = 'rounded',
  size = 'large',
  color = 'default',
  children,
  htmlType,
  ...rest
}: ButtonProps) => {
  return (
    <button
      className={classNames(
        'bgm-button',
        className,
        { 'bgm-button--disabled': disabled },
        `bgm-button--${type}`,
        {
          [`bgm-button--shape-${shape}`]: shape !== 'rounded',
          [`bgm-button--size-${size}`]: size !== 'large',
          [`bgm-button--color-${color}`]: color !== 'default',
        },
      )}
      type={htmlType}
      disabled={disabled}
      {...rest}
    >
      {children}
    </button>
  );
};

export type ButtonLinkProps = LinkProps & ButtonCommonProps;

export const ButtonLink = ({
  children,
  className,
  type = 'primary',
  shape = 'rounded',
  size = 'large',
  color = 'default',
  ...props
}: ButtonLinkProps) => {
  return (
    <Typography.Link
      className={classNames('bgm-button', className, 'bgm-button--link', `bgm-button--${type}`, {
        [`bgm-button--shape-${shape}`]: shape !== 'rounded',
        [`bgm-button--size-${size}`]: size !== 'large',
        [`bgm-button--color-${color}`]: color !== 'default',
      })}
      noStyle
      {...props}
    >
      {children}
    </Typography.Link>
  );
};

Button.Link = ButtonLink;

export default Button;
