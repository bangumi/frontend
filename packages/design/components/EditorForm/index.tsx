import './style';

import classnames from 'classnames';
import React, { forwardRef } from 'react';

import Button from '../Button';
import Link from '../Typography/Link';
import type { EditorProps } from './Editor';
import Editor from './Editor';

export interface EditorFormProps extends EditorProps {
  /** 最外层 className */
  className?: string;
  /** 最外层 style */
  style?: React.CSSProperties;
  /** 确认按钮的文本 */
  confirmText?: string;
  /** 确认按钮后的回调 */
  onConfirm?: (content: string) => void;
  /** 取消按钮的文本 */
  cancelText?: string;
  /** 取消按钮的回调 */
  onCancel?: () => void;
  /**
   * 是否隐藏取消按钮
   * @default false
   */
  hideCancel?: boolean;
}

const EditorForm = forwardRef<HTMLTextAreaElement, EditorFormProps>(
  (
    {
      className,
      style,
      confirmText = '写好了',
      onConfirm,
      cancelText = '取消',
      onCancel,
      hideCancel = false,
      ...props
    },
    ref,
  ) => {
    return (
      <div className={classnames('bgm-editor__form', className)} style={style}>
        <Editor ref={ref} onConfirm={onConfirm} {...props} />
        <div className='bgm-editor__submit'>
          <Button
            color='blue'
            className='bgm-editor__button bgm-editor__button--confirm'
            onClick={() => onConfirm?.(props.value ?? '')}
          >
            {confirmText}
          </Button>
          {!hideCancel && (
            <Button type='text' className='bgm-editor__button' onClick={onCancel}>
              {cancelText}
            </Button>
          )}
          <span className='bgm-editor__bbcode-tip'>
            <span className='bgm-editor__bbcode-tip__left'>
              使用 Ctrl+Enter 或 Alt+S 快速提交 |{' '}
            </span>
            <Link isExternal to='https://bgm.tv/help/bbcode'>
              BBCode指南
            </Link>
          </span>
        </div>
      </div>
    );
  },
);

export default EditorForm;
