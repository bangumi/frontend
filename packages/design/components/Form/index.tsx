import './style';

import cn from 'classnames';
import type { PropsWithChildren, ReactNode } from 'react';
import React, { createContext, useContext } from 'react';

type ItemProps = PropsWithChildren<{
  label?: ReactNode;
}>;

function FormItem({ children, label }: ItemProps) {
  const { labelWidth, compact } = useFormContext();
  return (
    <div className='bgm-form-item'>
      {!compact && (
        <div
          className='bgm-form-item-label'
          style={{
            width: `${labelWidth}px`,
          }}
        >
          <label>{label}</label>
        </div>
      )}
      {children}
    </div>
  );
}

const FormContext = createContext<{ labelWidth: number; compact: boolean } | null>(null);

export const useFormContext = () => {
  const context = useContext(FormContext);
  if (!context) {
    throw new Error('useFormContext() may only be used in the context of a <Form> component.');
  }
  return context;
};

export const useInFormContext = () => {
  const context = useContext(FormContext);
  return context !== null;
};

type FormProps = PropsWithChildren<{
  labelWidth?: number;
  compact?: boolean;
}> &
  JSX.IntrinsicElements['form'];

const Form = ({
  children,
  labelWidth,
  className,
  onKeyDown,
  compact = false,
  ...rest
}: FormProps) => {
  return (
    <FormContext.Provider value={{ labelWidth: labelWidth ?? 12, compact }}>
      <form
        className={cn('bgm-form', { 'bgm-form--compact': compact }, className)}
        onKeyDown={onKeyDown}
        {...rest}
      >
        {children}
      </form>
    </FormContext.Provider>
  );
};

Form.Item = FormItem;

export default Form;
