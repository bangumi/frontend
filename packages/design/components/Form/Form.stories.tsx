import './Form.stories.less';

import type { ComponentMeta, ComponentStory } from '@storybook/react';
import React, { useState } from 'react';
import type { SubmitHandler } from 'react-hook-form';
import { useForm } from 'react-hook-form';

import Button from '../Button';
import EditorForm from '../EditorForm';
import Input from '../Input';
import Select from '../Select';
import Form from '.';

const componentMeta: ComponentMeta<typeof Form> = {
  title: 'Modern/Form',
  component: Form,
  argTypes: {
    onSubmit: {
      action: 'summit',
    },
  },
};

export default componentMeta;

const Template: ComponentStory<typeof Form> = (args) => {
  const innerOnSubmit: SubmitHandler<{}> = (data) => {
    console.log(data);
  };
  const { register, handleSubmit } = useForm();

  return (
    <Form
      labelWidth={120}
      onSubmit={(e) => {
        args.onSubmit?.(e);
        handleSubmit(innerOnSubmit)(e);
      }}
    >
      <Form.Item label='类别名'>
        <Input type='text' {...register('name')} />
      </Form.Item>

      <Form.Item label='类型'>
        <Input type='text' {...register('type')} />
      </Form.Item>

      <Form.Item label='剧情介绍'>
        <textarea {...register('剧情')} />
      </Form.Item>

      <Form.Item label='编辑摘要'>
        <Input.Group>
          <Select
            options={[
              { label: '1', value: '1' },
              { label: '2', value: '2' },
              { label: '3', value: '3' },
            ]}
            style={{
              width: '120px',
            }}
            defaultValue='1'
          />
          <Input />
        </Input.Group>
      </Form.Item>

      <Form.Item label='描述信息'>
        <Input.Group>
          <Input
            placeholder='项目名'
            wrapperStyle={{
              width: '20%',
            }}
          />
          <Input />
        </Input.Group>
      </Form.Item>

      <Button htmlType='submit'>提交修改</Button>
    </Form>
  );
};

export const Default = Template.bind({});

export const Compact: ComponentStory<typeof Form> = (args) => {
  const [content, setContent] = useState('');

  return (
    <Form compact style={{ width: 675 }} {...args}>
      <Input type='text' placeholder='给新帖取一个标题' />
      <EditorForm
        placeholder='想聊点什么的呢…'
        value={content}
        onChange={setContent}
        confirmText='快速发帖'
      />
    </Form>
  );
};

export const CompactWithInputGroup = () => {
  return (
    <Form compact style={{ maxWidth: 675 }}>
      <Form.Item>
        <Input type='text' placeholder='给新帖取一个标题' />
      </Form.Item>
      <Form.Item>
        <Input.Group>
          <Input
            placeholder='项目名'
            wrapperStyle={{
              width: '20%',
            }}
          />
          <Input />
          <Input />
        </Input.Group>
      </Form.Item>
      <Form.Item>
        <Input type='text' placeholder='给新帖取一个标题' />
      </Form.Item>
      <Form.Item>
        <Input.Group>
          <Input
            placeholder='项目名'
            wrapperStyle={{
              width: '20%',
            }}
          />
          <Input />
        </Input.Group>
      </Form.Item>
    </Form>
  );
};
