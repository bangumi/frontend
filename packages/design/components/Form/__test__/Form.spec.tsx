import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import Button from '../../Button';
import EditorForm from '../../EditorForm';
import Input from '../../Input';
import Form from '..';

describe('Form Components', () => {
  it('should submit when submit button clicked', async () => {
    const handleSubmit = vi.fn();
    const { findByTestId } = render(
      <Form
        onSubmit={(e) => {
          e.preventDefault();
          handleSubmit();
        }}
      >
        <Button htmlType='submit' data-testid='submit'>
          Submit
        </Button>
      </Form>,
    );
    const btn = await findByTestId('submit');
    btn.click();
    await waitFor(() => {
      expect(handleSubmit).toBeCalled();
    });
  });

  it('key down should work properly', async () => {
    const onKeyDown = vi.fn();
    const { getByTestId } = render(
      <Form data-testid='form' onKeyDown={onKeyDown}>
        <Input id='input' data-testid='input' />
        <Button htmlType='submit' data-testid='submit'>
          Submit
        </Button>
      </Form>,
    );
    const form = getByTestId('form');
    fireEvent.keyDown(form, { key: 'Enter', code: 'Enter' });
    await waitFor(() => {
      expect(onKeyDown).toBeCalled();
    });
  });

  it('form item render properly', () => {
    const { container } = render(
      <Form labelWidth={120}>
        <Form.Item>
          <Input />
        </Form.Item>
        <Form.Item>
          <Input.Group>
            <Input />
            <Input />
          </Input.Group>
        </Form.Item>
      </Form>,
    );
    expect(container).toMatchSnapshot();
  });

  it('render compact layout properly', () => {
    const { container } = render(
      <Form compact>
        <Input />
        <Input />
        <EditorForm />
      </Form>,
    );
    expect(container).toMatchSnapshot();
  });
});
