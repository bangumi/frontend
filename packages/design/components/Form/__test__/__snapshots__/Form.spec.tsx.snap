// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form Components > form item render properly 1`] = `
<div>
  <form
    class="bgm-form"
  >
    <div
      class="bgm-form-item"
    >
      <div
        class="bgm-form-item-label"
        style="width: 120px;"
      >
        <label />
      </div>
      <div
        class="bgm-input__wrapper"
      >
        <input
          class="bgm-input"
          type="text"
        />
      </div>
    </div>
    <div
      class="bgm-form-item"
    >
      <div
        class="bgm-form-item-label"
        style="width: 120px;"
      >
        <label />
      </div>
      <div
        class="bgm-input-group"
      >
        <div
          class="bgm-input__wrapper"
        >
          <input
            class="bgm-input"
            type="text"
          />
        </div>
        <div
          class="bgm-input__wrapper"
        >
          <input
            class="bgm-input"
            type="text"
          />
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Form Components > render compact layout properly 1`] = `
<div>
  <form
    class="bgm-form bgm-form--compact"
  >
    <div
      class="bgm-input__wrapper"
    >
      <input
        class="bgm-input"
        type="text"
      />
    </div>
    <div
      class="bgm-input__wrapper"
    >
      <input
        class="bgm-input"
        type="text"
      />
    </div>
    <div
      class="bgm-editor__form"
    >
      <div
        class="bgm-editor__container"
      >
        <div
          class="bgm-editor__toolbar"
        >
          <div
            class="bgm-editor__toolbox"
          >
            <div
              data-testid="bold"
              title="粗体 [Ctrl+B]"
            />
            <div
              data-testid="italic"
              title="斜体 [Ctrl+I]"
            />
            <div
              data-testid="underscore"
              title="下划线 [Ctrl+U]"
            />
            <div
              data-testid="image"
              title="图片 [Ctrl+P]"
            />
            <div
              data-testid="link"
              title="链接 [Ctrl+L]"
            />
            <div
              data-testid="size"
              title="文字尺寸 [Ctrl+S]"
            />
          </div>
        </div>
        <textarea
          class="bgm-editor__text"
        />
      </div>
      <div
        class="bgm-editor__submit"
      >
        <button
          class="bgm-button bgm-editor__button bgm-editor__button--confirm bgm-button--primary bgm-button--color-blue"
        >
          写好了
        </button>
        <button
          class="bgm-button bgm-editor__button bgm-button--text"
        >
          取消
        </button>
        <span
          class="bgm-editor__bbcode-tip"
        >
          <span
            class="bgm-editor__bbcode-tip__left"
          >
            使用 Ctrl+Enter 或 Alt+S 快速提交 |
             
          </span>
          <a
            class="bgm-link"
            href="https://bgm.tv/help/bbcode"
          >
            BBCode指南
          </a>
        </span>
      </div>
    </div>
  </form>
</div>
`;
